src/tools/schema2graph.ts

```diff
@@ -0,0 +1,144 @@
+/**
+ * @license
+ * Copyright (c) 2019 Google Inc. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * Code distributed by Google as part of this project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+import {Schema} from '../runtime/schema.js';
+import {Direction} from '../runtime/manifest-ast-nodes.js';
+import {Dictionary} from '../runtime/hot.js';
+import {Manifest} from '../runtime/manifest.js';
+
+export interface FieldEntry {
+  particleName: string;
+  schema: Schema;
+  direction: Direction;
+  connectionName: string;
+}
+
+export interface TypeLattice {
+  nodes: Dictionary<FieldEntry>;
+  edges: Dictionary<string[]>;
+  nodesByParticle: Dictionary<FieldEntry[]>;
+  contains(name: string): boolean;
+}
+
+class Builder {
+  manifestData: Iterable<FieldEntry> = null;
+
+  public from(manifest: Manifest): Builder {
+    this.manifestData = this.parse(manifest);
+    return this;
+  }
+
+  public to(graph: TypeGraph) {
+    if (this.manifestData === null) {
+      throw new Error('Call `from` on a manifest first!');
+    }
+    this.buildGraph(this.manifestData, graph);
+  }
+
+  public build(): TypeLattice {
+    const graph = new TypeGraph();
+    this.to(graph);
+    return graph;
+  }
+
+  protected* parse(manifest: Manifest): Iterable<FieldEntry> {
+    for (const particle of manifest.allParticles) {
+      for (const connection of particle.connections) {
+        const schema: Schema = connection.type.getEntitySchema();
+        if (!schema) {
+          continue;
+        }
+        yield {
+          particleName: particle.name,
+          schema,
+          direction: connection.direction,
+          connectionName: connection.name,
```

Comment:
Looks like this could immediately encode the 'Foo_handle' name, rather then re-create that each time `nameFromEntry` is called.